subsys/mpsl/mpsl_init.c

```diff
@@ -118,6 +121,125 @@ static uint8_t m_config_clock_source_get(void)
 #endif
 }
 
+#if CONFIG_MPSL_FEM_NRF21540_GPIO
+static int fem_nrf21540_gpio_configure(void)
+{
+	/* FEM configuration requires gpiote and ppi channels.
+	 * Currently there is no reliable common method to dynamically
+	 * allocate such channels. FEM module needs only "some" channels
+	 * to use whichever they are, but FEM needs them for exclusive use
+	 * and does not enable them immediately.
+	 *
+	 * When common api to assign gpiote and ppi channels is available
+	 * current solution based on macros coming from Kconfig should be
+	 * reworked.
+	 */
+
+#if !DT_NODE_EXISTS(DT_NODELABEL(nrf_radio_fem))
+#error Node with label 'nrf_radio_fem' not found in the devicetree.
+#endif
+
+	mpsl_fem_nrf21540_gpio_interface_config_t cfg = {
+		.fem_config = {
+			.pa_time_gap_us  =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), tx_en_settle_time_us),
+			.lna_time_gap_us =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), rx_en_settle_time_us),
+			.pdn_settle_us   =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), pdn_settle_time_us),
+			.trx_hold_us     =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), trx_hold_time_us),
+			.pa_gain_db      =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), tx_gain_db),
+			.lna_gain_db     =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), rx_gain_db)
+		},
+		.pa_pin_config = {
+			.enable       = true,
+			.active_high  = true,
+			.gpio_pin     = DT_PROP(DT_NODELABEL(nrf_radio_fem), tx_en_pin),
+			.gpiote_ch_id = CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_TX_EN
+		},
+		.lna_pin_config = {
+			.enable       = true,
+			.active_high  = true,
+			.gpio_pin     = DT_PROP(DT_NODELABEL(nrf_radio_fem), rx_en_pin),
+			.gpiote_ch_id = CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_RX_EN
+		},
+		.pdn_pin_config = {
+			.enable       = true,
+			.active_high  = true,
+			.gpio_pin     = DT_PROP(DT_NODELABEL(nrf_radio_fem), pdn_pin),
+			.gpiote_ch_id = CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_PDN
+		},
+		.ppi_channels = {
+			CONFIG_MPSL_FEM_NRF21540_GPIO_PPI_CHANNEL_0,
+			CONFIG_MPSL_FEM_NRF21540_GPIO_PPI_CHANNEL_1,
+			CONFIG_MPSL_FEM_NRF21540_GPIO_PPI_CHANNEL_2
+		}
+	};
+
+	return mpsl_fem_nrf21540_gpio_interface_config_set(&cfg);
+}
+#endif
+
+#if CONFIG_MPSL_FEM_SKY66112_11
+static int fem_sky66112_11_configure(void)
+{
+#if !DT_NODE_EXISTS(DT_NODELABEL(nrf_radio_fem))
+#error Node with label 'nrf_radio_fem' not found in the devicetree.
+#endif
+
+	mpsl_fem_simple_gpio_interface_config_t cfg = {
+		.fem_config = {
+			.pa_time_gap_us  =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), ctx_settle_time_us),
+			.lna_time_gap_us =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), crx_settle_time_us),
+			.pa_gain_db      =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), tx_gain_db),
+			.lna_gain_db     =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), rx_gain_db)
+		},
+		.pa_pin_config = {
+			.enable       = true,
```

Comment:
In case of nRF21540 GPIO interface we are targeting one specific FEM (from Nordic, nRF21540), so it is not meant to be 'generic' or something, which is not the case in case of 'Simple GPIO' FEM. nRF21540 has a negative PDN polarity, so we this is exactly what we provide.

Reply:
Are there cases were the user only wants to enable the PA or only the LNA? If not, I'm fine with the configuration

Reply:
Good suggestion. It may be advantageous if these two parameters are also configurable with device-tree. I will add one more commit with this

Reply:
Now it is possible to set the polarity and also to disable pins